File: app/src/main/java/com/topjohnwu/magisk/utils/PatchAPK.java
package com.topjohnwu.magisk.utils;

import android.content.ComponentName;
import android.widget.Toast;

import androidx.core.app.NotificationCompat;

import com.topjohnwu.magisk.App;
import com.topjohnwu.magisk.BuildConfig;
import com.topjohnwu.magisk.ClassMap;
import com.topjohnwu.magisk.Config;
import com.topjohnwu.magisk.Const;
import com.topjohnwu.magisk.R;
import com.topjohnwu.magisk.SplashActivity;
import com.topjohnwu.magisk.uicomponents.Notifications;
import com.topjohnwu.signing.JarMap;
import com.topjohnwu.signing.SignAPK;
import com.topjohnwu.superuser.Shell;

import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.CharBuffer;
import java.nio.IntBuffer;
import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.List;
import java.util.jar.JarEntry;

public class PatchAPK {

    public static final String LOWERALPHA = "abcdefghijklmnopqrstuvwxyz";
    public static final String UPPERALPHA = LOWERALPHA.toUpperCase();
    public static final String ALPHA = LOWERALPHA + UPPERALPHA;
    public static final String DIGITS = "0123456789";
    public static final String ALPHANUM = ALPHA + DIGITS;
    public static final String ALPHANUMDOTS = ALPHANUM + "............";

    private static String genPackageName(String prefix, int length) {
        StringBuilder builder = new StringBuilder(length);
        builder.append(prefix);
        length -= prefix.length();
        SecureRandom random = new SecureRandom();
        char next, prev = prefix.charAt(prefix.length() - 1);
        for (int i = 0; i < length; ++i) {
            if (prev == '.' || i == length - 1) {
                next = ALPHA.charAt(random.nextInt(ALPHA.length()));
            } else {
                next = ALPHANUMDOTS.charAt(random.nextInt(ALPHANUMDOTS.length()));
            }
            builder.append(next);
            prev = next;
        }
        return builder.toString();
    }

    private static boolean findAndPatch(byte xml[], String from, String to) {
        if (from.length() != to.length())
            return false;
        CharBuffer buf = ByteBuffer.wrap(xml).order(ByteOrder.LITTLE_ENDIAN).asCharBuffer();
        List<Integer> offList = new ArrayList<>();
        for (int i = 0; i < buf.length() - from.length(); ++i) {
            boolean match = true;
            for (int j = 0; j < from.length(); ++j) {
                if (buf.get(i + j) != from.charAt(j)) {
                    match = false;
                    break;
                }
            }
            if (match) {
                offList.add(i);
                i += from.length();
            }
        }
        if (offList.isEmpty())
            return false;
        for (int off : offList) {
            buf.position(off);
            buf.put(to);
        }
        return true;
    }

    private static boolean findAndPatch(byte xml[], int a, int b) {
        IntBuffer buf = ByteBuffer.wrap(xml).order(ByteOrder.LITTLE_ENDIAN).asIntBuffer();
        int len = xml.length / 4;
        for (int i = 0; i < len; ++i) {
            if (buf.get(i) == a) {
                buf.put(i, b);
                return true;
            }
        }
        return false;
    }

    private static boolean patchAndHide() {
        App app = App.self;

        // Generate a new app with random package name
        File repack = new File(app.getFilesDir(), "patched.apk");
        String pkg = genPackageName("com.", BuildConfig.APPLICATION_ID.length());

        if (!patch(app.getPackageCodePath(), repack.getPath(), pkg))
            return false;

        // Install the application
        repack.setReadable(true, false);
        if (!Shell.su("pm install " + repack).exec().isSuccess())
            return false;

        Config.set(Config.Key.SU_MANAGER, pkg);
        Config.export();
        RootUtils.rmAndLaunch(BuildConfig.APPLICATION_ID,
                new ComponentName(pkg, ClassMap.get(SplashActivity.class).getName()));

        return true;
    }

    public static boolean patch(String in, String out, String pkg) {
        try {
            JarMap jar = new JarMap(in);
            JarEntry je = jar.getJarEntry(Const.ANDROID_MANIFEST);
            byte xml[] = jar.getRawData(je);

            if (!findAndPatch(xml, BuildConfig.APPLICATION_ID, pkg) ||
                    !findAndPatch(xml, R.string.app_name, R.string.re_app_name))
                return false;

            // Write in changes
            jar.getOutputStream(je).write(xml);
            SignAPK.sign(jar, new BufferedOutputStream(new FileOutputStream(out)));
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
        return true;
    }

    public static void hideManager() {
        App.THREAD_POOL.execute(() -> {
            App app = App.self;
            NotificationCompat.Builder progress =
                    Notifications.progress(app.getString(R.string.hide_manager_title));
            Notifications.mgr.notify(Const.ID.HIDE_MANAGER_NOTIFICATION_ID, progress.build());
            if(!patchAndHide())
                Utils.toast(R.string.hide_manager_fail_toast, Toast.LENGTH_LONG);
            Notifications.mgr.cancel(Const.ID.HIDE_MANAGER_NOTIFICATION_ID);
        });
    }
}

File: app/src/main/java/com/topjohnwu/magisk/utils/PatchAPK.kt
package com.topjohnwu.magisk.utils

import android.content.Context
import android.os.Build.VERSION.SDK_INT
import android.widget.Toast
import com.topjohnwu.magisk.*
import com.topjohnwu.magisk.data.network.GithubRawServices
import com.topjohnwu.magisk.extensions.DynamicClassLoader
import com.topjohnwu.magisk.extensions.get
import com.topjohnwu.magisk.extensions.subscribeK
import com.topjohnwu.magisk.extensions.writeTo
import com.topjohnwu.magisk.view.Notifications
import com.topjohnwu.signing.JarMap
import com.topjohnwu.signing.SignAPK
import com.topjohnwu.superuser.Shell
import io.reactivex.Completable
import timber.log.Timber
import java.io.File
import java.io.FileOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.SecureRandom

object PatchAPK {

    private const val LOWERALPHA = "abcdefghijklmnopqrstuvwxyz"
    private val UPPERALPHA = LOWERALPHA.toUpperCase()
    private val ALPHA = LOWERALPHA + UPPERALPHA
    private const val DIGITS = "0123456789"
    private val ALPHANUM = ALPHA + DIGITS
    private val ALPHANUMDOTS = "$ALPHANUM............"

    private fun genPackageName(prefix: String, length: Int): String {
        val builder = StringBuilder(length)
        builder.append(prefix)
        val len = length - prefix.length
        val random = SecureRandom()
        var next: Char
        var prev = prefix[prefix.length - 1]
        for (i in 0 until len) {
            next = if (prev == '.' || i == len - 1) {
                ALPHA[random.nextInt(ALPHA.length)]
            } else {
                ALPHANUMDOTS[random.nextInt(ALPHANUMDOTS.length)]
            }
            builder.append(next)
            prev = next
        }
        return builder.toString()
    }

    private fun findAndPatch(xml: ByteArray, from: String, to: String): Boolean {
        if (to.length > from.length)
            return false
        val buf = ByteBuffer.wrap(xml).order(ByteOrder.LITTLE_ENDIAN).asCharBuffer()
        val offList = mutableListOf<Int>()
        var i = 0
        loop@ while (i < buf.length - from.length) {
            for (j in from.indices) {
                if (buf.get(i + j) != from[j]) {
                    ++i
                    continue@loop
                }
            }
            offList.add(i)
            i += from.length
        }
        if (offList.isEmpty())
            return false

        val toBuf = to.toCharArray().copyOf(from.length)
        for (off in offList) {
            buf.position(off)
            buf.put(toBuf)
        }
        return true
    }

    private fun patchAndHide(context: Context, label: String): Boolean {
        // If not running as stub, and we are compatible with stub, use stub
        val src = if (!isRunningAsStub && SDK_INT >= 28 && Info.env.connectionMode == 3) {
            val stub = File(context.cacheDir, "stub.apk")
            val svc = get<GithubRawServices>()
            runCatching {
                svc.fetchFile(Info.remote.stub.link).blockingGet().byteStream().use {
                    it.writeTo(stub)
                }
            }.onFailure {
                Timber.e(it)
                return false
            }
            stub.path
        } else {
            context.packageCodePath
        }

        // Generate a new random package name and signature
        val repack = File(context.cacheDir, "patched.apk")
        val pkg = genPackageName("com.", BuildConfig.APPLICATION_ID.length)
        Config.keyStoreRaw = ""

        if (!patch(src, repack.path, pkg, label))
            return false

        // Install the application
        repack.setReadable(true, false)
        if (!Shell.su("force_pm_install $repack").exec().isSuccess)
            return false

        Config.suManager = pkg
        Config.export()
        Shell.su("pm uninstall ${BuildConfig.APPLICATION_ID}").submit()

        return true
    }

    @JvmStatic
    @JvmOverloads
    fun patch(apk: String, out: String, pkg: String, label: String = "Manager"): Boolean {
        try {
            val jar = JarMap.open(apk)
            val je = jar.getJarEntry(Const.ANDROID_MANIFEST)
            val xml = jar.getRawData(je)

            if (!findAndPatch(xml, BuildConfig.APPLICATION_ID, pkg) ||
                    !findAndPatch(xml, "Magisk Manager", label))
                return false

            // Write apk changes
            jar.getOutputStream(je).write(xml)
            val keys = Keygen()
            SignAPK.sign(keys.cert, keys.key, jar, FileOutputStream(out).buffered())
        } catch (e: Exception) {
            Timber.e(e)
            return false
        }

        return true
    }

    fun patch(apk: File, out: File, pkg: String, label: String): Boolean {
        try {
            if (apk.length() < 1 shl 18) {
                // APK is smaller than 256K, must be stub
                return patch(apk.path, out.path, pkg, label)
            }

            // Try using the new APK to patch itself
            val loader = DynamicClassLoader(apk)
            val cls = loader.loadClass("a.a")

            for (m in cls.declaredMethods) {
                val pars = m.parameterTypes
                if (pars.size == 4 && pars[0] == String::class.java) {
                    return m.invoke(null, apk.path, out.path, pkg, label) as Boolean
                }
            }
            throw Exception("No matching method found")
        } catch (e: Exception) {
            Timber.e(e)
            // Fallback to use the current implementation
            return patch(apk.path, out.path, pkg, label)
        }
    }

    fun hideManager(context: Context, label: String) {
        Completable.fromAction {
            val progress = Notifications.progress(context, context.getString(R.string.hide_manager_title))
            Notifications.mgr.notify(Const.ID.HIDE_MANAGER_NOTIFICATION_ID, progress.build())
            if (!patchAndHide(context, label))
                Utils.toast(R.string.hide_manager_fail_toast, Toast.LENGTH_LONG)
            Notifications.mgr.cancel(Const.ID.HIDE_MANAGER_NOTIFICATION_ID)
        }.subscribeK()
    }
}
